kernel/Utilities/SparsityEstimator.cpp

```diff
@@ -15,310 +18,416 @@ SparsityEstimator::SparsityEstimator(const GlobalIndexing& rowindexing,
+                        owned[rowInterval].emplace(colInterval);
+                    }
+                    for (Detail::DoFInterval colInterval :
+                         nonOwnedColIntervals) {
+                        nonOwned[rowInterval].emplace(colInterval);
+                    }
+                }
             }
         }
-        // Store the information
-        writeDoFCount(face, workspace, nonZeroPerRowOwned,
-                      nonZeroPerRowNonOwned);
     }
-    logger(VERBOSE, "Sparsity pattern for Face DoFs computed");
-    // Edges
-    for (const Base::Edge* edge : rowIndexing_.getMesh()->getEdgesList()) {
-        if (!edge->isOwnedByCurrentProcessor() ||
-            !hasLocalDoFs(edge, rowIndexing_.getIncludedUnknowns())) {
-            continue;
-        }
-        workspace.clear();
-        for (const Base::Element* element : edge->getElements()) {
-            // DoFs from supporting element
-            addElementDoFs(element, workspace);
-            // Face matrix coupling with the element on the other side of the
-            // face
-            if (!includeFaceCoupling) continue;
-            // See remark for face based DoFs
-            logger.assert_always(
-                element->isOwnedByCurrentProcessor(),
-                "Face coupling with conforming basis functions on distributed "
-                "meshes is not supported");
-            for (const Base::Face* face : element->getFacesList()) {
-                if (!face->isInternal()) continue;
-                const Base::Element* otherElement =
-                    face->getPtrOtherElement(element);
-                addElementDoFs(otherElement, workspace);
-            }
+
+    // If there is no face coupling anywhere (e.g. dg mass matrix) then we can
+    // skip looping over the adjacent elements.
+    bool anyFaceCoupling = false;
+    for (std::size_t i = 0; i < faceCoupling.getSize(); ++i) {
+        if (faceCoupling[i]) {
+            anyFaceCoupling = true;
+            break;
         }
-        // Store the information
-        writeDoFCount(edge, workspace, nonZeroPerRowOwned,
-                      nonZeroPerRowNonOwned);
     }
-    logger(VERBOSE, "Sparsity pattern for Edge DoFs computed");
-    // Nodes
-    if (rowIndexing_.getMesh()->dimension() > 1) {
-        for (const Base::Node* node : rowIndexing_.getMesh()->getNodesList()) {
-            if (!node->isOwnedByCurrentProcessor() ||
-                !hasLocalDoFs(node, rowIndexing_.getIncludedUnknowns())) {
-                continue;
+
+    bool distributedMesh = false;
+
+    if (anyFaceCoupling) {
+        for (const Base::Face* face : rowIndexing_.getMesh()->getFacesList()) {
+            // Check for a distributed mesh
+            if (face->getFaceType() == Geometry::FaceType::SUBDOMAIN_BOUNDARY ||
+                face->getFaceType() ==
+                    Geometry::FaceType::PERIODIC_SUBDOMAIN_BC) {
+                distributedMesh = true;
             }
```

Comment:
can we turn this into a small separate function, like `return bool CheckdistributedMesh()` ?

Reply:
Could but with a horrible complexity. As it would currently need to traverse all faces of the mesh.

I think it may be more useful to have a function to check if a face is a subdomain boundary. I'll add that.